File: day7/part2.py
from typing import List
from collections import defaultdict


def read_input(path: str) -> List[str]:
    with open(path, "r", encoding="utf-8") as f:
        return [line.rstrip("\n") for line in f if line.strip()]


def count_timelines(grid: List[str]) -> int:
    rows, cols = len(grid), len(grid[0])
    start_col = grid[0].index("S")

    dp = defaultdict(int)
    dp[(0, start_col)] = 1

    total = 0

    for r in range(rows):
        next_dp = defaultdict(int)

        for (row, col), count in dp.items():
            if row != r:
                continue

            if col < 0 or col >= cols:
                total += count
                continue

            cell = grid[row][col]

            if row == rows - 1:
                total += count
                continue

            if cell == "^":
                next_dp[(row + 1, col - 1)] += count
                next_dp[(row + 1, col + 1)] += count
            else:
                next_dp[(row + 1, col)] += count

        dp = next_dp

    return total


if __name__ == "__main__":
    grid = read_input("input.txt")
    print(count_timelines(grid))

File: day7/part2.ts
import * as fs from "fs";

type Key = string;

function readInput(path: string): string[] {
  return fs
    .readFileSync(path, "utf8")
    .split("\n")
    .map((l) => l.trim())
    .filter((l) => l.length > 0);
}

function key(r: number, c: number): Key {
  return `${r},${c}`;
}

function countTimeLines(grid: string[]): bigint {
  const rows = grid.length;
  const cols = grid[0].length;
  const startCol = grid[0].indexOf("S");

  let dp = new Map<Key, bigint>();
  dp.set(key(0, startCol), 1n);

  let total = 0n;

  for (let r = 0; r < rows; r++) {
    const next = new Map<Key, bigint>();

    for (const [k, count] of dp) {
      const [row, col] = k.split(",").map(Number);

      if (row !== r) continue;

      if (col < 0 || col >= cols || row === rows - 1) {
        total += count;
        continue;
      }

      if (grid[row][col] === "^") {
        next.set(
          key(row + 1, col - 1),
          (next.get(key(row + 1, col - 1)) ?? 0n) + count,
        );
        next.set(
          key(row + 1, col + 1),
          (next.get(key(row + 1, col + 1)) ?? 0n) + count,
        );
      } else {
        next.set(
          key(row + 1, col),
          (next.get(key(row + 1, col)) ?? 0n) + count,
        );
      }
    }

    dp = next;
  }

  return total;
}

const grid = readInput("input.txt");
console.log(countTimeLines(grid).toString());
